app/scripts/metamask-controller.js

```diff
@@ -401,9 +406,49 @@ module.exports = class MetamaskController extends EventEmitter {
       // notices
       checkNotices: noticeController.updateNoticesList.bind(noticeController),
       markNoticeRead: noticeController.markNoticeRead.bind(noticeController),
+
+      approveWeb3Request: this.approveWeb3Request.bind(this),
+      rejectWeb3Request: this.rejectWeb3Request.bind(this),
     }
   }
 
+  /**
+   * Called when a tab requests web3 access
+   *
+   * @param {string} requestedOrigin - Origin of the window requesting web3 access
+   */
+  handleWeb3Request (requestedOrigin) {
+    const { openPopup } = this.opts
+    this.memStore.updateState({ pendingWeb3Requests: [requestedOrigin] })
```

Comment:
Probably worth moving into its own controller long-term. Non-blocking, this works for MVP.